### ISSUE TITLE
installhtml: many uninitialized value warnings

### ISSUE THREAD

**Issue by jkeenan** [Perl/perl5#19052]
On or about July 25, @sisyphus reported on the results of applying a patch and using `installhtml` to install documentation in HTML format:

> However, with Pod-Html-1.31, I see the following warning repeated many times:
> 
> ```
> Use of uninitialized value $Podroot in string eq at C:/perl-5.35.2-1110/lib/Pod/Html.pm line 372.
> ```
> 
> This warning, which is absent with Pod-Html-1.28, occurs both when installing perl, and when subsequently installing some modules from the CPAN (eg Win32::IPC) into perl.
> 

I was able to reproduce these uninitialized value warnings but anticipated that they would get cleared up in an upcoming pull release.  I believe that that will indeed happen once https://github.com/Perl/perl5/pull/19050 is merged to blead.  I am creating this issue so that we can properly log progress.

**Comment by sisyphus** [comment]
Just to confirm that  multiple occurrences of essentially the same warning still appear when running 'make install' on perl-5.35.3 (released earlier today), though the line number has now changed:
```
Use of uninitialized value $Podroot in string eq at ..\lib/Pod/Html.pm line 431.
```
I realize that it's anticipated that this warning won't disappear until #19050 is merged - which is something that has not yet happened.

Other than that, 'make install' is now working fine for me with perl-5.35.3 on Windows 7, gcc-11.1.0.

**Comment by jkeenan** [comment]
On 8/21/21 8:38 AM, sisyphus wrote:
> Just to confirm that multiple occurrences of essentially the same 
> warning still appear when running 'make install' on perl-5.35.3 
> (released earlier today), though the line number has now changed:
> 
> |Use of uninitialized value $Podroot in string eq at ..\lib/Pod/Html.pm 
> line 431. |
> 
> I realize that it's anticipated that this warning won't disappear until 
> #19050 <https://github.com/Perl/perl5/pull/19050> is merged - which is 
> something that has not yet happened.
> 

Would it be possible for you to build a perl from p.r. 19050, attempt 
'installhtml' and see whether that actually does eliminate the warnings?

It would be better to clear this problem up before the merge rather than 
after.

> Other than that, 'make install' is now working fine for me with 
> perl-5.35.3 on Windows 7, gcc-11.1.0.
> 

Thanks.

**Comment by sisyphus** [comment]
On Sat, Aug 21, 2021 at 10:53 PM James E Keenan ***@***.***>
wrote:

> Would it be possible for you to build a perl from p.r. 19050, attempt
> 'installhtml' and see whether that actually does eliminate the warnings?
>
>
I see that #19050 has now been merged into blead.
So I've just tested using current blead (commit 3fd2ac55c) and can report
that the warnings I reported are no longer occurring.

However, I now notice some other warnings (near the end of 'make install').
I've just checked, and these warnings were also present in the devel
release (5.35.3) that was made on 21.08.
I haven't checked to see whether they were also present on 5.35.2 - but I
could check on that if it helps.
Here is how 'make install' now (successfully) ends:

.....
..\perl.exe -I..\lib ..\installhtml --podroot=.. --htmldir=.\html \
    --podpath=pod:lib:utils --htmlroot="file://C|\perl-5.35.4-1110\html"\
    --recurse
..\installhtml: Warning both '../lib/ExtUtils/ParseXS.pod' and
'../lib/ExtUtils/ParseXS.pm' exist, using pod
Use of uninitialized value $str in substitution (s///) at
..\lib/Pod/Simple/XHTML.pm line 75.
..\installhtml: Warning both '../lib/Locale/Maketext.pod' and
'../lib/Locale/Maketext.pm' exist, using pod
..\installhtml: Warning both '../lib/Module/CoreList.pod' and
'../lib/Module/CoreList.pm' exist, using pod
..\installhtml: Warning both '../lib/Pod/Simple.pod' and
'../lib/Pod/Simple.pm'exist, using pod
..\installhtml: Warning both '../lib/Config.pod' and '../lib/Config.pm'
exist, using pod
..\installhtml: Warning both '../lib/perlfaq.pod' and '../lib/perlfaq.pm'
exist, using pod
..\installhtml: Warning both '../lib/POSIX.pod' and '../lib/POSIX.pm'
exist, using pod
..\installhtml: Warning both '../lib/version.pod' and '../lib/version.pm'
exist, using pod
Use of uninitialized value $str in substitution (s///) at
..\lib/Pod/Simple/XHTML.pm line 75.
xcopy /f /r /i /e /d /y .\html\*.* C:\perl-5.35.4-1110\html\
C:\comp-1110\perl-5.35.4\win32\html\lib\Pod\Html.html ->
C:\perl-5.35.4-1110\html\html\lib\Pod\Html.html
C:\comp-1110\perl-5.35.4\win32\html\lib\Pod\Simple\HTML.html ->
C:\perl-5.35.4-1110\html\html\lib\Pod\Simple\HTML.html
C:\comp-1110\perl-5.35.4\win32\html\lib\Pod\Simple\HTMLBatch.html ->
C:\perl-5.35.4-1110\html\html\lib\Pod\Simple\HTMLBatch.html
C:\comp-1110\perl-5.35.4\win32\html\lib\Pod\Simple\HTMLLegacy.html ->
C:\perl-5.35.4-1110\html\html\lib\Pod\Simple\HTMLLegacy.html
4 File(s) copied

Let me know if you'd like me to investigate this further.

Cheers,
Rob

**Comment by jkeenan** [comment]
@sisyphus wrote,

> However, I now notice some other warnings (near the end of 'make install'). I've just checked, and these warnings were also present in the devel release (5.35.3) that was made on 21.08. I haven't checked to see whether they were also present on 5.35.2 - but I could check on that if it helps.

Warnings like this:
```
installhtml: Warning both '../lib/Locale/Maketext.pod' and '../lib/Locale/Maketext.pm' exist, using pod
```
... are being generated by `installhtml` itself and are to be expected.

As for warnings like this:
```
Use of uninitialized value $str in substitution (s///) at ..\lib/Pod/Simple/XHTML.pm line 75. 
```
I don't think these warnings represent any defect in either `installhtml` or `ext/Pod-Html/*`.  Rather, I think they are defects in particular modules containing POD.

Suppose you were to run `installhtml` in "slurp mode," *i.e.,* with `$|=1;`.  (See https://github.com/jkeenan/perl5/tree/installhtml-slurp-20210823.)  Suppose you were to run it with an invocation like this:
```
export HTMLDIR=$HOMEDIR/tmp/installhtml && \
export PODROOTDIR=$HOMEDIR/gitwork/perl/ && \
export PODPATH='cpan:ext:lib:pod:utils' && \
./perl -Ilib ./installhtml    \
            --podroot=$PODROOTDIR     \
            --podpath=$PODPATH \
            --htmldir=$HTMLDIR     \
            --htmlroot=$HTMLDIR    \
            --splithead=pod/perlipc     \
            --recurse \
            --verbose 2>&1 | tee $P5P_DIR/pod-html/installhtml.log
```
... with appropriate modifications for the envvars.

In installhtml.log you could grep for output like this:
```
$ zgrep -n -C1 "uninitialized value" installhtml.8102fae30a.log.gz
341-/home/jkeenan/gitwork/perl//cpan/Win32/Win32.pm => /home/jkeenan/tmp/installhtml/cpan/Win32/Win32.html
342:Use of uninitialized value $str in substitution (s///) at lib/Pod/Simple/XHTML.pm line 75.
343-/home/jkeenan/gitwork/perl//cpan/Math-BigInt/lib/Math/BigInt/Lib.pm => /home/jkeenan/tmp/installhtml/cpan/Math-BigInt/lib/Math/BigInt/Lib.html
--
550-/home/jkeenan/gitwork/perl//dist/IO/lib/IO/Handle.pm => /home/jkeenan/tmp/installhtml/dist/IO/lib/IO/Handle.html
551:Use of uninitialized value $str in substitution (s///) at lib/Pod/Simple/XHTML.pm line 75.
552-/home/jkeenan/gitwork/perl//dist/IO/IO.pm => /home/jkeenan/tmp/installhtml/dist/IO/IO.html
--
839-/home/jkeenan/gitwork/perl//lib/IO/Handle.pm => /home/jkeenan/tmp/installhtml/lib/IO/Handle.html
840:Use of uninitialized value $str in substitution (s///) at lib/Pod/Simple/XHTML.pm line 75.
841-/home/jkeenan/gitwork/perl//lib/IO/Pipe.pm => /home/jkeenan/tmp/installhtml/lib/IO/Pipe.html
```
I suspect the problems lie within `cpan/Win32/Win32.pm` and `dist/IO/lib/IO/Handle.pm`.  (I ran this on FreeBSD, so modules under `cpan/Win32` are not copied to `lib`.  Hence, the error appears in association with Win32.pm only once.)

If you manipulate the assignment to `$PODPATH` by adding or subtracting directories, you can see that the uninitialized value warnings disappear if, for example, you only call for `ext`.

Can you verify?

**Comment by jkeenan** [comment]
@sisyphus I'll be closing this issue within 7 days unless you indicate that it should be kept open.

Thank you very much.
Jim Keenan

**Comment by sisyphus** [comment]
On Thu, Sep 2, 2021 at 8:01 AM James E Keenan ***@***.***>
wrote:

> @sisyphus <https://github.com/sisyphus> I'll be closing this issue within
> 7 days unless you indicate that it should be kept open.
>

I think it can be closed.
I agree that the problem likely lies in the POD rather than installhtml
and/or Pod-Html, though I haven't found the time to actually verify that.

The main thing for me is that 'make install'  runs to completion - and that
is now happening.
Thanks for fixing that.
